wurst/network/Network.wurst

```diff
@@ -0,0 +1,922 @@
+		if getIndexFromChar(getCharFromIndex(i)) != i
+			print("CollisionTest failed for " + i.toString() + " " + getCharFromIndex(i))
+			print("getIndexFromChar() == " + getIndexFromChar(getCharFromIndex(i)).toString())
+			print("getCharFromIndex(getIndexFromChar)) == " + getCharFromIndex(getIndexFromChar(getCharFromIndex(i))))
+
+function getKey(integer i) returns string
+	if i >= gcKeysCount
+		error("Network: trying to get invalid GC key")
+	return gcKeys[i]
+
+function getCharFromIndex(integer i) returns string
+	if i > STRING_ALPHABET.length()
+		error("Network: trying to get invalid character from index '" + i.toString() + "'")
+	return STRING_ALPHABET.charAt(i)
+
+function getIndexFromChar(string c) returns integer
+	integer hash
+	if (c.toUpperCase() == c and c.toLowerCase() != c) or c == "/"
+		hash = ("00" + c).getHash()
+	else
+		hash = c.getHash()
+
+	if SAFETY_CHECKS_ENABLED and not alphabetLookup.hasInt(hash)
+		error("Network: trying to get index from invalid character '" + c + "'")
+
+	return alphabetLookup.loadInt(hash)
+
+init
+	for i = 1 to MAX_CHARACTERS_PER_KEY - 1
+		execute(() -> precomputeKeys(i))
+	execute(() -> precomputeLookup())
+	execute(() -> testCollisions())
+
+// simple box class to allow modifying a single value inside a closure
+class Box<T>
+	T value
+
+	construct(T value)
+		this.value = value
+
+class BoxBoolean
+	boolean value
+
+	construct(boolean value)
+		this.value = value
+
+class BoxString
+	string value
+
+	construct(string value)
+		this.value = value
+
+/**
+	Interface to be implemented by classes that wish to be directly
+	read/written to Network
+*/
+public interface Serializable
+	function serialize(DataStream stream)
+	function deserialize(DataStream stream)
+
+enum StreamType
+	INTEGER
+	REAL
+	BOOLEAN
+
+enum StreamMode
+	WRITING
+	READING
+	LOCKED
+
+/**
+	This class is used as an intermediate buffer to store data before it is synchronized.
+	That way, we can send data through gamecache using multiple bursts, while keeping the
+	gamecache key lengths short.
+**/
+public class DataStream
+	// max characters to store per compacting round
+	private static constant integer CHARS_PER_COMPACT = 512
+
+	private Table buffer = new Table()
+	// stores strings as int-sequences
+	private Table stringBuffer = new Table()
+
+	private StreamMode currentMode
+
+	private integer integerWriteIndex = -1
+	private integer realWriteIndex    = -1
+	private integer booleanWriteIndex = -1
+	private integer stringWriteIndex  = -1
+	private integer stringIntWriteIndex = -1
+
+	private integer integerReadIndex = -1
+	private integer realReadIndex    = -1
+	private integer booleanReadIndex = -1
+	private integer stringReadIndex  = -1
+	private integer stringIntReadIndex = -1
+
+	private integer compactIndex = -1
+	private integer explodeIndex = -1
+
+
+	construct()
+		currentMode = WRITING
+
+	ondestroy
+		destroy buffer
+
+	private function checkMode(StreamMode toCheck)
+		if currentMode != toCheck
```

Comment:
put `if SAFETY_CHECKS_ENABLED` in here instead of wrapping each call

Reply:
Thanks